scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py

```diff
@@ -67,20 +66,44 @@ def _makeMatMat(m):
         return lambda v: m @ v
 
 
+def _dot_inplace(x, y, verbosityLevel=0):
+    """Perform 'np.dot' in-place if dtypes match."""
+    if x.flags["CARRAY"] and x.shape[1] == y.shape[1] and x.dtype == y.dtype:
```

Comment:
```suggestion
    if (x.flags["CARRAY"] and x.shape[1] == y.shape[1] and x.dtype == y.dtype
            and x.dtype != np.float32):
```

Reply:
Alternatively/additionally, we could try some more flags here like: `x.flags['C_CONTIGUOUS'] and x.flags['WRITEABLE']`

Reply:
"CARRAY" includes x.flags['C_CONTIGUOUS'] and x.flags['WRITEABLE'], see https://numpy.org/doc/stable/reference/generated/numpy.ndarray.flags.html

Reply:
I suspect that float32 by itself is not a problem. I could of course exclude float32 as you propose, if no actual reason if found for the failure in time.